modules/lang-mustache/src/main/java/org/elasticsearch/script/mustache/RestSearchTemplateAction.java

```diff
@@ -74,6 +74,7 @@ public RestChannelConsumer prepareRequest(RestRequest request, NodeClient client
         SearchTemplateRequest searchTemplateRequest;
         try (XContentParser parser = request.contentOrSourceParamParser()) {
             searchTemplateRequest = SearchTemplateRequest.fromXContent(parser);
+            searchTemplateRequest.setExplain(request.paramAsBoolean("explain", searchTemplateRequest.isExplain()));
```

Comment:
While this fix seems to do the same job, it makes me wonder why we need to have special treatment for `explain` and whether we have the same problems with other flags like `simulate` and `profile`. It seems to me that `explain` is properly read into the inner search request, but the flag is also replicated into the top-level SearchTemplateRequest and the two end up having a different value. Do you see what I mean?

Reply:
> While this fix seems to do the same job, it makes me wonder why we need to have special treatment for `explain` and whether we have the same problems with other flags like `simulate` and `profile`

It doesn't seem to be a problem for the flags `simulate` and `profile`, because they are not specified as query parameters. When reading the file `search-template-api.asciidoc` , I can't see them described in the session `search-template-api-query-params`. Also, when trying to specify them using some rest client like Postman, I get `illegal_argument_exception.

```
{
    "error": {
        "root_cause": [
            {
                "type": "illegal_argument_exception",
                "reason": "request [/index03/_search/template] contains unrecognized parameter: [profile]"
            }
        ],
        "type": "illegal_argument_exception",
        "reason": "request [/index03/_search/template] contains unrecognized parameter: [profile]"
    },
    "status": 400
}
```




> It seems to me that `explain` is properly read into the inner search request, but the flag is also replicated into the top-level SearchTemplateRequest and the two end up having a different value.

I am assuming that
- top level request is related to the class SearchRequest
- inner search request is related to the class SearchTemplateRequest

I couldn't find any reference for the explain flag in SearchRequest.

I could find references for the explain flag in SearchTemplateRequest. I understand that every property was set only by this method SearchTemplateRequest.fromXContent(parser). Because of that, the only way to set that variable was using the body of the request. With the new line of code that I added, the query param will be considered as well.

Please let me know if I could get the right understanding of this problem. If I couldn't, please give some more inputs, and I'll get into the right track to address this issue. Tks @javanna

Reply:
thanks for digging ;)

You are right about `profile` and `simulate`, they are only accepted as part of the request body.

The explain flag is set as part of `SearchSourceBuilder` which is inside the `SearchRequest`, set in `RestSearchAction#parseSearchSource`. Looking at the code, it may be already read correctly without your change, but set to a different place, which makes me wonder why we duplicated it in the top-level SearchTemplateRequest (tangentially related, profile and simulate are also duplicated, but that's a different problem, let's leave that aside).

Your fix does make it work, but I am wondering if we'd need to do some additional cleanup on this and/or instead rely on the parsing that already happens within search source, provided that it does work like I think it does. Could you double check that?

Reply:
Hi @javanna 

Sorry for the delayed response. I was really busy at my work, but I do believe I can put more effort into this issue from now on. 

I could get your point, and I will dive deeper to try to understand why the flags are duplicated. 

By now, I believe that the simplest solution would be to set the `explain()` method to read both values, so the transport request would always get the right value. But I will try to figure out a way to avoid this duplicated flag and get a more elegant solution.

So far, thanks for your careful review!

Reply:
sounds good, thanks. Sorry that while this is a potentially straight-forward change, it reveals a potential clean-up :)